### PR TITLE
Add error message for failed SSO authorization

### LINKED ISSUE #6409: Process callback error given when user doesn't have a matching role in Teleport
### Description

**What happened**:

Users get the generic message unable to process callback when their traits don't match to roles.  This can create the impression that something is wrong with Teleport when it's typically the configuration.

image

**What you expected to happen**:

Teleport in most cases has processed the callback but there is not a matching role. This message could be given that would give the user the context that they did authenticate with their SSO but they do not have a matching role in Teleport configured.  that's different then when an actual callback error occurs that there is no connectivity or other technical issue.

### Reproduction Steps
As minimally and precisely as possible, describe step-by-step how to reproduce the problem.
1. Configure a Github or other SSO connector.  
2. Log in with a user that has a team or group that doesn't match to a available Teleport role
3.The error message will show.

### Server Details
- Teleport version (run `teleport version`): 6.1.0
- Server OS (e.g. from `/etc/os-release`): ubuntu

### PR DESCRIPTION
**Pull request by atburke**
This PR, along with gravitational/webapps#530, adds a new error message for when a user successfully authenticates with SSO but fails to match any roles.

Resolves #6409.

### CODE REVIEW

**Review by gabrielcorado — APPROVED** [review]

File: lib/client/redirect.go [hunk 1]
```diff
@@ -42,6 +42,10 @@ const (
 	// LoginFailedBadCallbackRedirectURL is a redirect URL when an SSO error specific to
 	// auth connector's callback was encountered.
 	LoginFailedBadCallbackRedirectURL = "/web/msg/error/login/callback"
+
+	// LoginFailedUnauthorizedRedirectURL is a redirect URL for when an SSO authenticates successfully,
+	// but the user has no matching roles in Teleport.
+	LoginFailedUnauthorizedRedirectURL = "/web/msg/error/login/auth"
```
Comment (codingllama) [inline lib/client/redirect.go:48]: Maybe:

```suggestion
	LoginFailedUnauthorizedRedirectURL = "/web/msg/error/login/noroles"
```

File: lib/web/apiserver_test.go [hunk 2]
```diff
@@ -3565,3 +3566,35 @@ type mockProxySettings struct {
...
 	return &webclient.ProxySettings{}, nil
 }
+
+func TestIsRoleMatchError(t *testing.T) {
+	t.Parallel()
+	matches := []struct {
+		name string
+		err  error
+	}{
+		{"saml", trace.AccessDenied("unable to map attributes to role for connector: x")},
```
Comment (codingllama) [inline lib/web/apiserver_test.go:3576]: nit: use keyed struct initialization

```suggestion
		{name: "saml", err: trace.AccessDenied("unable to map attributes to role for connector: x")},
```

same for others.

File: lib/web/apiserver_test.go [hunk 3]
```diff
@@ -3565,3 +3566,35 @@ type mockProxySettings struct {
...
+		t.Run(fmt.Sprintf("Accept %s", tc.name), func(t *testing.T) {
+			require.True(t, isRoleMatchError(tc.err))
+		})
+	}
+
+	nonMatches := []struct {
+		name string
+		err  error
+	}{
+		{"nil", nil},
```
Comment (codingllama) [inline lib/web/apiserver_test.go:3590]: nit: use keyed struct initialization

```suggestion
		{name: "nil"},
```

same for others.

File: lib/web/apiserver.go [hunk 4]
```diff
@@ -1056,6 +1056,9 @@ func (h *Handler) githubCallback(w http.ResponseWriter, r *http.Request, p httpr
 	response, err := h.cfg.ProxyClient.ValidateGithubAuthCallback(r.URL.Query())
 	if err != nil {
 		logger.WithError(err).Error("Error while processing callback.")
+		if isRoleMatchError(err) {
```
Comment (codingllama) [inline lib/web/apiserver.go:1059]: Any chance we could the change to this callback, in addition to isRoleMatchError?

Same for `oidcCallback` below.

File: lib/web/apiserver.go [hunk 5]
```diff
@@ -2701,3 +2707,18 @@ func ssoSetWebSessionAndRedirectURL(w http.ResponseWriter, r *http.Request, resp
 
 	return nil
 }
+
+// isRoleMatchError checks if an error is the result of a failed role match during
+// SSO authorization.
+func isRoleMatchError(err error) bool {
+	// oidc or saml
+	if trace.IsAccessDenied(err) {
+		return strings.HasPrefix(err.Error(), "unable to map attributes to role") ||
```
Comment (codingllama) [inline lib/web/apiserver.go:2716]: Should we do Contains checks here instead, in case some data is added prepended to the error before this is called?

**Review by codingllama — COMMENTED** [review]
Revert the webassets changes?

File: lib/client/redirect.go (hunk 1 above)
Comment (atburke) [reply lib/client/redirect.go:48]: In the frontend, the new error message is `You are not authorized, please contact your SSO administrator.` I figured that there might be other situations where we want to show this message, and tying the URL to roles would be awkward.

File: lib/web/apiserver.go (hunk 4 above)
Comment (atburke) [reply lib/web/apiserver.go:1059]: I think you lost a word! Can you clarify this?

Comment (codingllama) [reply lib/web/apiserver.go:1059]: Sorry, I meant if we could test the change to the callback, in addition to the tests you added for isRoleMatchError.

File: lib/client/redirect.go (hunk 1 above)
Comment (codingllama) [reply lib/client/redirect.go:48]: Fair enough.

File: lib/web/apiserver.go [hunk 6]
```diff
@@ -2701,3 +2707,21 @@ func ssoSetWebSessionAndRedirectURL(w http.ResponseWriter, r *http.Request, resp
 
 	return nil
 }
+
+// isRoleMatchError checks if an error is the result of a failed role match during
+// SSO authorization.
+func isRoleMatchError(err error) bool {
+	// oidc or saml
+	if trace.IsAccessDenied(err) {
+		// from lib/auth/saml.go:calculateSAMLUser
```
Comment (codingllama) [inline lib/web/apiserver.go:2716]: I thought we didn't control the error strings, but if they are part of our codebase then let's ditch the comments and make sure they stay in sync (through a constant string, constant error, error.Is check, etc).

**Comment by zmb3** [comment]
@atburke any plans to push this through?

**Comment by atburke** [comment]
@zmb3 Eventually. It's just been a lower priority than my other work for a while.

File: lib/auth/oidc.go [hunk 7]
```diff
@@ -49,7 +50,7 @@ const (
 
 // IsOIDCNoRolesError checks if an error results from a failure to map roles from claims.
 func IsOIDCNoRolesError(err error) bool {
-	return trace.IsAccessDenied(err) && err.Error() == oidcNoRolesErrorMessage
+	return trace.IsAccessDenied(err) && strings.Contains(err.Error(), oidcNoRolesErrorMessage)
```
Comment (zmb3) [inline lib/auth/oidc.go:53]: Can we use `errors.Is` instead of relying on string matching here?

File: lib/auth/saml.go [hunk 8]
```diff
@@ -41,14 +40,7 @@ import (
...
 
-const (
-	samlNoRolesErrorMessage = "No roles mapped from claims. The mappings may contain typos."
-)
-
-// IsSAMLNoRolesError checks if an error results from a failure to map roles from claims.
-func IsSAMLNoRolesError(err error) bool {
-	return trace.IsAccessDenied(err) && strings.Contains(err.Error(), samlNoRolesErrorMessage)
-}
+var SAMLNoRolesError = trace.AccessDenied("No roles mapped from claims. The mappings may contain typos.")
```
Comment (zmb3) [inline lib/auth/saml.go:43]: Same here - do you need to export this?

File: lib/auth/oidc.go [hunk 9]
```diff
@@ -44,14 +43,7 @@ import (
...
 
-const (
-	oidcNoRolesErrorMessage = "No roles mapped from claims. The mappings may contain typos."
-)
-
-// IsOIDCNoRolesError checks if an error results from a failure to map roles from claims.
-func IsOIDCNoRolesError(err error) bool {
-	return trace.IsAccessDenied(err) && strings.Contains(err.Error(), oidcNoRolesErrorMessage)
-}
+var OIDCNoRolesError = trace.AccessDenied("No roles mapped from claims. The mappings may contain typos.")
```
Comment (zmb3) [inline lib/auth/oidc.go:46]: Does this need to be exported? It's safer if we can leave it unexported - this way other packages can't reassign it to something else.

**Comment by atburke** [comment]
@codingllama and/or @zmb3 Mind taking another look at this?

File: lib/auth/oidc_test.go [hunk 10]
```diff
@@ -186,143 +186,174 @@ func TestUserInfoBadStatus(t *testing.T) {
...
+				Identity: types.ExternalIdentity{
+					ConnectorID: "-sso-test-okta",
+					Username:    "superuser@example.com",
+				},
+				Req: *request,
+			}, resp)
+
+			diagCtx := ssoDiagContext{}
+
+			resp, err = s.a.validateOIDCAuthCallback(ctx, &diagCtx, values)
```
Comment (codingllama) [inline lib/auth/oidc_test.go:293]: Isn't this called as part of the public function above? Why do we go through it again?

**Review by codingllama — COMMENTED** [review]
Apologies for the delay!

I'm not super familiar with SSO impls, so the tests are specially challenging to parse. I suggest we CC someone comfortable with those parts of the codebase to review as well, as I'm afraid some of the finer points might go over my head.

File: lib/auth/oidc_test.go (hunk 10 above)
Comment (atburke) [reply lib/auth/oidc_test.go:293]: Not sure. The double validate was already part of the test before this PR.

**Review by codingllama — COMMENTED** [review]
Error handling is much better, thanks.

**Review by codingllama — APPROVED** [review]

**Pull request merged**